test/integration/eds_integration_test.cc

```diff
@@ -437,5 +451,471 @@ TEST_P(EdsIntegrationTest, StatsReadyFilter) {
+    const auto& hostset_per_priority = cluster_ref.get().prioritySet().hostSetsPerPriority();
+    EXPECT_EQ(2, hostset_per_priority.size());
+    const Envoy::Upstream::HostSetPtr& host_set = hostset_per_priority[0];
+    EXPECT_EQ(expected_factor, host_set->overprovisioningFactor());
+  };
+  get_and_compare(Envoy::Upstream::kDefaultOverProvisioningFactor);
+  // Use new overprovisioning factor 200.
+  setEndpoints(4, 4, 0, true, 200);
+  get_and_compare(200);
+}
+
+// Verifies that EDS update only triggers member update callbacks once per update.
+TEST_P(EdsOverGrpcIntegrationTest, BatchMemberUpdateCb) {
+  initializeTest(false);
+  uint32_t member_update_count{};
+  auto& priority_set = test_server_->server()
+                           .clusterManager()
+                           .clusters()
+                           .active_clusters_.find("cluster_0")
+                           ->second.get()
+                           .prioritySet();
+  // Keep track of how many times we're seeing a member update callback.
+  auto member_update_cb = priority_set.addMemberUpdateCb([&](const auto& hosts_added, const auto&) {
+    // We should see both hosts present in the member update callback.
+    EXPECT_EQ(2, hosts_added.size());
+    member_update_count++;
+  });
+  setEndpoints(2, 2, 0);
+  EXPECT_EQ(1, member_update_count);
+}
+
+TEST_P(EdsOverGrpcIntegrationTest, StatsReadyFilter) {
+  config_helper_.prependFilter("name: eds-ready-filter");
+  initializeTest(false);
+  // Initial state: no healthy endpoints
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("500", response->headers().getStatusValue());
+  EXPECT_EQ("EDS not ready", response->body());
+  cleanupUpstreamAndDownstream();
+  // 2/2 healthy endpoints.
+  setEndpoints(2, 2, 0);
+  EXPECT_EQ(2, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  response = IntegrationUtil::makeSingleRequest(lookupPort("http"), "GET", "/cluster1", "",
+                                                downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  EXPECT_EQ("EDS is ready", response->body());
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(EdsOverGrpcIntegrationTest, ReuseMuxAndStreamForMultipleClusters) {
+  initializeTest(false, 3, false);
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_total")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_1.membership_total")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_2.membership_total")->value());
+  switch (clientType()) {
+  case Grpc::ClientType::EnvoyGrpc:
```

Comment:
@adisuissa figured out the approach to verify the number of streams for both Grpc client types. For Envoy Grpc number of total streams towards specific cluster equals to total number of H2 requests. Google Grpc actually exposes stream metrics per client instance, which is nice. Please let me know wdyt.